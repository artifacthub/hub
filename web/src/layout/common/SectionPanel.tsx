import classnames from 'classnames';
import isUndefined from 'lodash/isUndefined';
import React, { useState } from 'react';

import { SectionItem } from '../../utils/data';
import styles from './SectionPanel.module.css';

interface Props {
  defaultSection?: number;
  sections: SectionItem[];
  content: {
    [key: string]: JSX.Element;
  };
}

const SectionPanel = (props: Props) => {
  const [activeSection, setActiveSection] = useState<number>(props.defaultSection || 0);

  const onMenuItemClick = (section: number) => {
    setActiveSection(section);
  };

  return (
    <main role="main" className="container d-flex flex-column flex-md-row justify-content-between my-md-4 p-0">
      <nav className={styles.sidebar}>
        <div className={`list-group my-4 my-md-0 mr-md-5 ${styles.listGroup}`}>
          {props.sections.map((section: SectionItem) => {
            return (
              <button
                data-testid="sectionBtn"
                key={`package_${section.index}`}
                type="button"
                className={classnames(
                  'list-group-item list-group-item-action d-flex flex-row align-items-center',
                  styles.listItem,
                  { [styles.isActive]: section.index === activeSection },
                  { disabled: section.disabled }
                )}
                disabled={section.disabled}
                onClick={() => onMenuItemClick(section.index)}
              >
                <div className="d-flex flex-row align-items-center">
                  {!isUndefined(section.icon) && <div className={styles.icon}>{section.icon}</div>}
                  <div className="ml-2">{section.name}</div>
                </div>
              </button>
            );
          })}
        </div>
      </nav>

      <div className={`flex-grow-1 ${styles.list}`}>{props.content[activeSection]}</div>
    </main>
  );
};

export default SectionPanel;
