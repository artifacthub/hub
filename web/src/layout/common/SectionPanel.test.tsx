import { fireEvent, render } from '@testing-library/react';
import React from 'react';

import SectionPanel from './SectionPanel';

const defaultProps = {
  sections: [
    { index: 0, name: 'Details 0', shortName: 'details_0', disabled: false },
    { index: 1, name: 'Details 1', shortName: 'details_1', disabled: false },
    { index: 2, name: 'Details 2', shortName: 'details_2', disabled: true },
  ],
  content: {
    0: <>Content 0</>,
    1: <>Content 1</>,
    2: <>Content 2</>,
  },
};

describe('SectionPanel', () => {
  it('creates snapshot', () => {
    const { asFragment } = render(<SectionPanel {...defaultProps} />);
    expect(asFragment).toMatchSnapshot();
  });

  it('renders proper content', () => {
    const { getAllByTestId, getByText } = render(<SectionPanel {...defaultProps} />);

    const btns = getAllByTestId('sectionBtn');

    expect(btns).toHaveLength(3);
    expect(btns[0]).toHaveTextContent(defaultProps.sections[0].name);
    expect(btns[1]).toHaveTextContent(defaultProps.sections[1].name);
    expect(btns[2]).toHaveTextContent(defaultProps.sections[2].name);
    expect(btns[2]).toBeDisabled();

    expect(getByText('Content 0')).toBeInTheDocument();
  });

  it('changes active section', () => {
    const { getAllByTestId, getByText } = render(<SectionPanel {...defaultProps} />);

    const btns = getAllByTestId('sectionBtn');
    fireEvent.click(btns[1]);

    expect(getByText('Content 1')).toBeInTheDocument();
  });

  it('renders with active section', () => {
    const { getByText } = render(<SectionPanel {...defaultProps} defaultSection={2} />);

    expect(getByText('Content 2')).toBeInTheDocument();
  });
});
