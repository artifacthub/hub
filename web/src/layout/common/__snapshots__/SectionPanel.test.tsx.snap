// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SectionPanel creates snapshot 1`] = `[Function]`;
